librz/bin/format/elf/elf.c

```diff
@@ -712,3604 +687,453 @@ static void fill_dynamic_entries(ELFOBJ *bin, ut64 loaded_offset, ut64 dyn_size)
 
-	rz_return_val_if_fail(bin, false);
-	if (!bin->phdr || !bin->ehdr.e_phnum) {
+	Elf_(Phdr) *dynamic = get_dynamic_segment(bin);
+	if (!dynamic) {
 		return false;
 	}
 
-	Elf_(Phdr) *dyn_phdr = get_dynamic_segment(bin);
-	if (!dyn_phdr) {
+	Elf_(Xword) dynamic_size = dynamic->p_filesz;
+	ut64 dynamic_offset = Elf_(rz_bin_elf_v2p_new)(bin, dynamic->p_vaddr);
+	if (dynamic_offset == UT64_MAX) {
 		return false;
 	}
 
-	dyn_size = dyn_phdr->p_filesz;
-	loaded_offset = Elf_(rz_bin_elf_v2p_new)(bin, dyn_phdr->p_vaddr);
-	if (loaded_offset == UT64_MAX) {
+	if (!dynamic_size || dynamic_offset + dynamic_size > bin->size) {
 		return false;
 	}
 
-	if (!dyn_size || loaded_offset + dyn_size > bin->size) {
-		return false;
-	}
+	fill_dynamic_entries(bin, dynamic_offset, dynamic_size);
+	init_dynamic_section_sdb(bin);
 
-	fill_dynamic_entries(bin, loaded_offset, dyn_size);
+	return true;
+}
 
-	if (bin->dyn_info.dt_strtab != RZ_BIN_ELF_ADDR_MAX) {
-		strtabaddr = Elf_(rz_bin_elf_v2p_new)(bin, bin->dyn_info.dt_strtab);
-	}
+/// Parse NT_PRSTATUS note
+static void parse_note_prstatus(ELFOBJ *bin, RzBinElfNote *note, Elf_(Nhdr) * note_segment_header, ut64 offset) {
+	RzBinElfPrStatusLayout *layout = Elf_(rz_bin_elf_get_prstatus_layout)(bin);
 
-	if (bin->dyn_info.dt_strsz > 0) {
-		strsize = bin->dyn_info.dt_strsz;
+	if (!layout) {
+		eprintf("Fetching registers from core file not supported for this architecture.\n");
+		return;
 	}
 
-	if (strtabaddr == UT64_MAX || strtabaddr > bin->size || strsize > ST32_MAX || !strsize || strsize > bin->size || strtabaddr + strsize > bin->size) {
-		if (!strtabaddr) {
-			bprintf("DT_STRTAB not found or invalid\n");
-		}
-		return false;
-	}
-	strtab = (char *)calloc(1, strsize + 1);
-	if (!strtab) {
-		return false;
-	}
-	r = rz_buf_read_at(bin->b, strtabaddr, (ut8 *)strtab, strsize);
-	if (r < 1) {
-		free(strtab);
-		return false;
+	ut8 *buf = malloc(layout->regsize);
+	if (!buf) {
+		return;
 	}
 
-	bin->strtab = strtab;
-	bin->strtab_size = strsize;
-	init_dynamic_section_sdb(bin, strtabaddr, strsize);
-	return true;
-}
-
-static void note_fini(RzBinElfNote *note) {
-	switch (note->type) {
-	case NT_FILE:
-		for (size_t i = 0; i < note->file.files_count; i++) {
-			free(note->file.files[i].file);
-		}
-		free(note->file.files);
-		break;
-	case NT_PRSTATUS:
-		free(note->prstatus.regstate);
-		break;
+	if (rz_buf_read_at(bin->b, offset + layout->regdelta, buf, layout->regsize) != layout->regsize) {
+		free(buf);
+		bprintf("Cannot read register state from CORE file\n");
+		return;
 	}
+
+	note->prstatus.regstate_size = layout->regsize;
+	note->prstatus.regstate = buf;
 }
 
-static void note_segment_free(RzBinElfNoteSegment *seg) {
-	if (!seg) {
-		return;
-	}
-	if (seg->notes) {
-		for (size_t i = 0; i < seg->notes_count; i++) {
-			note_fini(&seg->notes[i]);
-		}
-		free(seg->notes);
-	}
-	free(seg);
+static void set_note_file(ELFOBJ *bin, RzBinElfNoteFile *file, ut64 *offset, const char *file_name) {
```

Comment:
All of these notes, at least the ones we parse, are specific to core files, so they can go into `elf_corefile.c`.

Reply:
I can add (tomorrow) a elf_notes.c file if you want.

Reply:
sounds good to me.